src/vehicle_gui_base.h

```diff
@@ -12,19 +12,53 @@
 #ifndef VEHICLE_GUI_BASE_H
 #define VEHICLE_GUI_BASE_H
 
+#include "core/smallvec_type.hpp"
+#include "date_type.h"
+#include "economy_type.h"
 #include "sortlist_type.h"
 #include "vehiclelist.h"
 #include "window_gui.h"
 #include "widgets/dropdown_type.h"
 
+#include <iterator>
+
 typedef GUIList<const Vehicle*> GUIVehicleList;
 
+struct GUIVehicleGroup {
+	VehicleList::const_iterator vehicles_begin;    ///< Pointer to beginning element of this vehicle group.
+	VehicleList::const_iterator vehicles_end;      ///< Pointer to past-the-end element of this vehicle group.
+	Money display_profit_this_year;                ///< Total profit for the vehicle group this year.
+	Money display_profit_last_year;                ///< Total profit for the vehicle group laste year.
+	Date age;                                      ///< Age in days of oldest vehicle in the group.
+	std::ptrdiff_t NumVehicles() const {
+		return std::distance(vehicles_begin, vehicles_end);
+	}
+	const Vehicle * GetSingleVehicle() const {
+		assert(NumVehicles() == 1);
+		return vehicles_begin[0];
+	}
+};
+
+typedef GUIList<GUIVehicleGroup> GUIVehicleGroupList;
+
 struct BaseVehicleListWindow : public Window {
-	GUIVehicleList vehicles;  ///< The list of vehicles
-	Listing *sorting;         ///< Pointer to the vehicle type related sorting.
-	byte unitnumber_digits;   ///< The number of digits of the highest unit number
+
+	enum GroupBy : byte {
+		GB_NONE,
+
+		GB_END,
+	};
+
+	GroupBy grouping;                         ///< How we want to group the list.
+	std::vector<const Vehicle*> vehicles;     ///< List of vehicles.  This is the buffer for vehgroups to point into.
```

Comment:
missing space before '*'
Looks like this line is replaced by VehicleList in the next commit, no reason why this couldn't be done here